Changing the type of reference count from Session to atomic int

markccchiang opened the pull request:
It tries to solve #978. The reference count from Session is increased when `OnMessageTask` is created to run in a different thread. It is decreased when the task is done and `OnMessageTask` is destructed. So the type of reference count is better to be atomic int to avoid unexpected behavior in animation.

markccchiang committed:
change the type of reference count from Session to atomic int

ajm-ska commented:
I'm sorry to report that the issue appears to still remain.

veggiesaurus commented:
I do think we should make this change anyway, even if it doesn't fix the issue. @robsimmonds do you agree?

markccchiang committed:
set timeout mechanism for animation unit test

markccchiang commented:
It seems two animation tasks can not run at the same time. In this unit test, it runs an animation task play forward and then another task playback. I added a busy waiting step between these two tasks. In a rare case, the backend can not generate all the data messages we expected. So the while loop can not be broken and it hangs here. I added a timeout mechanism to break this infinite loop. Now it looks running smoothly on Ming-Yi's Ubuntu computer. @ajm-asiaa could you please test it again?

ajm-ska reviewed (approved):
@markccchiang Nice work! That appears to have fixed the issue all right. I ran the test on Ubuntu with `--gtest_repeat=100` and they all ran successfully without hanging.

acdo2002 commented:
Yes, I also test it with --gtest_repeat=100 on my Ubuntu laptop, they are all passed!

veggiesaurus commented:
> It seems two animation tasks can not run at the same time. In this unit test, it runs an animation task play forward and then another task playback. I added a busy waiting step between these two tasks. In a rare case, the backend can not generate all the data messages we expected. So the while loop can not be broken and it hangs here. I added a timeout mechanism to break this infinite loop. Now it looks running smoothly on Ming-Yi's Ubuntu computer. @ajm-asiaa could you please test it again?

@markccchiang just to be sure, does the test still fail if it times out?

veggiesaurus commented:
these changes look fine to me, but I know @robsimmonds is working on fixing the underlying issue with handling animation messages in an unexpected order, so I suggest we merge this in, so as not to lose our CI functionality, and then revisit the issue when @robsimmonds makes his PR.

We can thenclose the associated issue and make a new, more relevant one (specific to the animation tasks).

markccchiang commented:
> @markccchiang just to be sure, does the test still fail if it times out?

Yes, it reports an error if times out. Since the number of response messages does not fit our expectations.
